Fix use of equality of value_sett::object_mapt

romainbrenguier opened the pull request:
This was removed recently in pull request diffblue/cbmc#4694 but it
seems necessary when compiling with clang which would complain about it
not being defined.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Use read() for value_sett::object_mapt equality

The explicit definition of object_mapt equality was removed recently in
pull request diffblue/cbmc#4694 but it seems necessary when compiling
with clang which would complain about it not being defined.

Should fix https://github.com/diffblue/cbmc/issues/4699

smowton reviewed (commented):
I assume this is a fix for https://github.com/diffblue/cbmc/issues/4699 ? If so, note that in the commit message and add a comment that this is a pointless specialisation of https://github.com/diffblue/cbmc/blob/develop/src/util/reference_counting.h#L185 which we supply to tolerate old broken compilers.

romainbrenguier commented:
@smowton 
> I assume this is a fix for #4699 ? If so, note that in the commit message and add a comment that this is a pointless specialisation of https://github.com/diffblue/cbmc/blob/develop/src/util/reference_counting.h#L185 which we supply to tolerate old broken compilers.

Yes, I have added that to the commit message. I also changed a bit the code so that we don't redefine the equality function.

smowton commented:
Shame to throw away the reference-equal case though? If you're going to inline `reference_countingt::operator==` then inline the whole thing

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8cd09b34285fa97ddeb918c18ed8ffb7ff152964).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113132139

romainbrenguier commented:
@smowton 
> Shame to throw away the reference-equal case though? If you're going to inline reference_countingt::operator== then inline the whole thing

this is now fixed

smowton reviewed (approved):
Add a comment noting that this is a copy of (link to `reference_countingt::operator==`) and noting why this was done

allredj commented:
We're fixing something for clang 3.8, but we're not compiling with clang 3.8 on CI. So this (or something similar) will probably break later and we'll waste more time. Should we not add another build on CI?

romainbrenguier commented:
@allredj 
> We're fixing something for clang 3.8, but we're not compiling with clang 3.8 on CI. So this (or something similar) will probably break later and we'll waste more time. Should we not add another build on CI?

pull request to add the CI step: https://github.com/diffblue/cbmc/pull/4718

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b27ceb9897dda5cd527e886cd58f70fea586ef0b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113404218
